### PR TITLE
Add function for verifying blocks in the context of the chain

### PR DESCRIPTION
**Pull request by edolstra**
Currently, this only verifies slot leaders and block dates.

### CODE REVIEW

File: cardano/src/block/verify_chain.rs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        }
+    }
+}
+
+pub fn verify_block_in_chain(
+    protocol_magic: ProtocolMagic,
+    chain_state: &mut ChainState,
+    block_hash: &HeaderHash,
+    blk: &Block) -> Result<(), Error>
+{
```
Comment (NicolasDP) [inline cardano/src/block/verify_chain.rs]: why not make this function a method of `ChainState`?

```rust
    pub fn verify_block(&mut self, &HeaderHash, &Block) -> Result<&mut Self, Error>;
```

File: cardano/src/block/verify_chain.rs [hunk 2]
```diff
@@ -0,0 +1,104 @@
+use block::*;
+use address;
+use config::{ProtocolMagic};
+
+pub struct ChainState {
+    pub prev_block: HeaderHash,
+    pub prev_date: Option<BlockDate>,
+    pub slot_leaders: Vec<address::StakeholderId>,
```
Comment (NicolasDP) [inline cardano/src/block/verify_chain.rs]: I would also add the expected `protocol_magic` here as it is a parameter of the blockchain and it must be consistent within the whole blockchain.

File: cardano/src/block/verify_chain.rs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+    res
+}
+
+fn do_verify(
+    protocol_magic: ProtocolMagic,
+    chain_state: &mut ChainState,
+    block_hash: &HeaderHash,
+    blk: &Block) -> Result<(), Error>
+{
```
Comment (NicolasDP) [inline cardano/src/block/verify_chain.rs]: you can also make this function a private member of the ChainState:

```rust
    fn do_verify(&mut self, &HeaderHash, &Block) -> Result<&mut Self, Error>;
```

File: cardano/src/block/verify_chain.rs (hunk 1 above)
Comment (NicolasDP) [reply cardano/src/block/verify_chain.rs]: The return of `&mut Self` is not necessary here.

**Comment by edolstra** [comment]
@NicolasDP Updated.

**Pull request merged**